Please mention how to view help for a specific task in 'usage' output

When `gradle` or `gradlew` is run with the name of a task and an option that the program does not understand, it may print out one of two things (see 'Current Behavior', below, for a description). I assume for internal or architectural reasons that it is not feasible for these two to be the same, so in this issue I ask for one small detail to be added to the first 'usage' variant.

### Expected Behavior
<!--- Tell us how it should work -->
The output shown when one of the gradle tool's built-in options is misused should include one extra message, as shown below:

```
$ ./gradlew wrapper --version=6.4.1-all

Command-line option '--version' does not take an argument.
* Try:
Run gradlew help --task wrapper to get task usage details

USAGE: gradlew [option...] [task...]

-?, -h, --help                     Shows this help message.
-a, --no-rebuild                   Do not rebuild project dependencies.
-b, --build-file                   Specify the build file.
# .... and so on
```

### Current Behavior
<!--- Explain the difference from current behavior -->
Running a command with an unsupported command-line option may lead to one of two outputs.

1. a plain message consisting of
    * an error message such as `Command-line option '--profile' does not take an argument.`
    * a long (and somewhat helpful) list of supported command-line options
2. a rich message with color and text styling consisting of
    * an all-caps **FAILURE: Build failed with an exception.**
    * an error message such as `Problem configuring task :my-task from command line.`
    * an additional detail message such as `> Unknown command-line option '--use'.`
    * a suggestion to run the `help` task with this as a parameter
    * a suggestion to view the gradle website for help
    * (optional) a link to some document page about deprecated warnings
    * (optional) advice on how to view more details about these specific warnings
    * a reminder that your build failed, in case you missed it
    * how long the process ran

Respective output:
<details><summary>Output with reserved option</summary>

```
$ ./gradlew wrapper --version=6.4.1-all

Command-line option '--version' does not take an argument.

USAGE: gradlew [option...] [task...]


-?, -h, --help                     Shows this help message.
-a, --no-rebuild                   Do not rebuild project dependencies.
-b, --build-file                   Specify the build file.
--build-cache                      Enables the Gradle build cache. Gradle will try to reuse outputs from previous builds.
-c, --settings-file                Specify the settings file.
--configure-on-demand              Configure necessary projects only. Gradle will attempt to reduce configuration time for large multi-project builds. [incubating]
--console                          Specifies which type of console output to generate. Values are 'plain', 'auto' (default), 'rich' or 'verbose'.
--continue                         Continue task execution after a task failure.
-D, --system-prop                  Set system property of the JVM (e.g. -Dmyprop=myvalue).
-d, --debug                        Log in debug mode (includes normal stacktrace).
--daemon                           Uses the Gradle Daemon to run the build. Starts the Daemon if not running.
--export-keys                      Exports the public keys used for dependency verification. [incubating]
-F, --dependency-verification      Configures the dependency verification mode (strict, lenient or off) [incubating]
--foreground                       Starts the Gradle Daemon in the foreground.
-g, --gradle-user-home             Specifies the gradle user home directory.
-I, --init-script                  Specify an initialization script.
-i, --info                         Set log level to info.
--include-build                    Include the specified build in the composite.
-M, --write-verification-metadata  Generates checksums for dependencies used in the project (comma-separated list) [incubating]
-m, --dry-run                      Run the builds with all task actions disabled.
--max-workers                      Configure the number of concurrent workers Gradle is allowed to use.
--no-build-cache                   Disables the Gradle build cache.
--no-configure-on-demand           Disables the use of configuration on demand. [incubating]
--no-daemon                        Do not use the Gradle daemon to run the build. Useful occasionally if you have configured Gradle to always run with the daemon by default.
--no-parallel                      Disables parallel execution to build projects.
--no-scan                          Disables the creation of a build scan. For more information about build scans, please visit https://gradle.com/build-scans.
--offline                          Execute the build without accessing network resources.
-P, --project-prop                 Set project property for the build script (e.g. -Pmyprop=myvalue).
-p, --project-dir                  Specifies the start directory for Gradle. Defaults to current directory.
--parallel                         Build projects in parallel. Gradle will attempt to determine the optimal number of executor threads to use.
--priority                         Specifies the scheduling priority for the Gradle daemon and all processes launched by it. Values are 'normal' (default) or 'low' [incubating]
--profile                          Profile build execution time and generates a report in the <build_dir>/reports/profile directory.
--project-cache-dir                Specify the project-specific cache directory. Defaults to .gradle in the root project directory.
-q, --quiet                        Log errors only.
--refresh-dependencies             Refresh the state of dependencies.
--refresh-keys                     Refresh the public keys used for dependency verification. [incubating]
--rerun-tasks                      Ignore previously cached task results.
-S, --full-stacktrace              Print out the full (very verbose) stacktrace for all exceptions.
-s, --stacktrace                   Print out the stacktrace for all exceptions.
--scan                             Creates a build scan. Gradle will emit a warning if the build scan plugin has not been applied. (https://gradle.com/build-scans)
--status                           Shows status of running and recently stopped Gradle Daemon(s).
--stop                             Stops the Gradle Daemon if it is running.
-t, --continuous                   Enables continuous build. Gradle does not exit and will re-execute tasks when task file inputs change.
--update-locks                     Perform a partial update of the dependency lock, letting passed in module notations change version. [incubating]
-v, --version                      Print version info.
-w, --warn                         Set log level to warn.
--warning-mode                     Specifies which mode of warnings to generate. Values are 'all', 'fail', 'summary'(default) or 'none'
--write-locks                      Persists dependency resolution for locked configurations, ignoring existing locking information if it exists [incubating]
-x, --exclude-task                 Specify a task to be excluded from execution.

```
</details>

<details><summary>Output with un-reserved option</summary>

```
$ ./gradlew wrapper --use=6.4.1-all

FAILURE: Build failed with an exception.

* What went wrong:
Problem configuring task :wrapper from command line.
> Unknown command-line option '--use'.

* Try:
Run gradlew help --task :wrapper to get task usage details. Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

Deprecated Gradle features were used in this build, making it incompatible with Gradle 7.0.
Use '--warning-mode all' to show the individual deprecation warnings.
See https://docs.gradle.org/6.4.1/userguide/command_line_interface.html#sec:command_line_warnings

BUILD FAILED in 1s
```
</details>

### Context
<!--- How has this issue affected you? What are you trying to accomplish? What other alternatives have you considered? -->
<!--- Providing context helps us come up with a solution that is most useful in the real world -->
I was using gradle 6.4.1 and intended to confirm that my local wrapper installation was up-to-date. In doing so, I wanted to try out [the command-line arguments offered by that task](https://docs.gradle.org/current/userguide/gradle_wrapper.html#sec:upgrading_wrapper), but could not remember the exact spelling off of the top of my head.

My first attempt (using a `--version=` option to specify which version of the wrapper I wanted) showed me the warning above. Curious, I tried running `gradlew --help wrapper` and `gradlew wrapper --help`, which were even less helpful. Lastly, I searched `wrapper` on the gradle documentation site, which ultimately led me to the page I linked in the prior paragraph. Only after experimenting with nonsense options did I find the message that would have helped: `Run gradlew help --task :wrapper to get task usage details`.

--- a/platforms/core-runtime/gradle-cli/src/main/java/org/gradle/launcher/cli/DefaultCommandLineActionFactory.java
+++ b/platforms/core-runtime/gradle-cli/src/main/java/org/gradle/launcher/cli/DefaultCommandLineActionFactory.java
@@ -94,8 +94,23 @@ public class DefaultCommandLineActionFactory implements CommandLineActionFactory
         return new DefaultBuildClientMetaData(new GradleLauncherMetaData());
     }
 
-    private static void showUsage(PrintStream out, CommandLineParser parser) {
+    private static void showUsage(PrintStream out, CommandLineParser parser, @Nullable String suggestedTaskSelector) {
         out.println();
+        out.print("To see more detail about a task, run ");
+        clientMetaData().describeCommand(out, "help --task <task>");
+        out.println();
+        if (suggestedTaskSelector != null) {
+            out.print("For example, ");
+            clientMetaData().describeCommand(out, "help --task " + suggestedTaskSelector);
+            out.println();
+        }
+        out.println();
+        out.print("To see a list of available tasks, run ");
+        clientMetaData().describeCommand(out, "tasks");
+        out.println();
+
+        out.println();
+
         out.print("USAGE: ");
         clientMetaData().describeCommand(out, "[option...]", "[task...]");
         out.println();
@@ -139,7 +154,7 @@ public class DefaultCommandLineActionFactory implements CommandLineActionFactory
         @Nullable
         public Action<? super ExecutionListener> createAction(CommandLineParser parser, ParsedCommandLine commandLine, Parameters parameters) {
             if (commandLine.hasOption(HELP)) {
-                return new ShowUsageAction(parser);
+                return new ShowUsageAction(parser, commandLine);
             }
             if (commandLine.hasOption(VERSION)) {
                 return new ShowVersionAction(parameters);
@@ -166,25 +181,40 @@ public class DefaultCommandLineActionFactory implements CommandLineActionFactory
         private final Exception exception;
         private final CommandLineParser parser;
 
-        public CommandLineParseFailureAction(CommandLineParser parser, Exception exception) {
+        private final List<String> args;
+
+        public CommandLineParseFailureAction(CommandLineParser parser, Exception exception, List<String> args) {
             this.parser = parser;
             this.exception = exception;
+            this.args = args;
         }
 
         @Override
         public void execute(ExecutionListener executionListener) {
             System.err.println();
             System.err.println(exception.getMessage());
-            showUsage(System.err, parser);
+            showUsage(System.err, parser, getSuggestedTaskSelector());
             executionListener.onFailure(exception);
+        }
+
+        @Nullable
+        private String getSuggestedTaskSelector() {
+            for (String arg : args) {
+                if (!arg.startsWith("-")) {
+                    return arg;
+                }
+            }
+            return null;
         }
     }
 
     private static class ShowUsageAction implements Action<ExecutionListener> {
         private final CommandLineParser parser;
+        private final ParsedCommandLine commandLine;
 
-        public ShowUsageAction(CommandLineParser parser) {
+        public ShowUsageAction(CommandLineParser parser, ParsedCommandLine commandLine) {
             this.parser = parser;
+            this.commandLine = commandLine;
         }
 
         @Override
@@ -193,7 +223,16 @@ public class DefaultCommandLineActionFactory implements CommandLineActionFactory
             System.out.print("To see help contextual to the project, use ");
             clientMetaData().describeCommand(System.out, "help");
             System.out.println();
-            showUsage(System.out, parser);
+            showUsage(System.out, parser, getSuggestedTaskSelector());
+        }
+
+        @Nullable
+        private String getSuggestedTaskSelector() {
+            if (!commandLine.getExtraArguments().isEmpty()) {
+                return commandLine.getExtraArguments().get(0);
+            } else {
+                return null;
+            }
         }
     }
 
@@ -325,7 +364,7 @@ public class DefaultCommandLineActionFactory implements CommandLineActionFactory
                 Parameters parameters = buildEnvironmentConfigurationConverter.convertParameters(commandLine, null);
                 action = createAction(parser, commandLine, parameters);
             } catch (CommandLineArgumentException e) {
-                action = new CommandLineParseFailureAction(parser, e);
+                action = new CommandLineParseFailureAction(parser, e, args);
             }
 
             action.execute(executionListener);

--- a/platforms/core-runtime/gradle-cli/src/test/groovy/org/gradle/launcher/cli/DefaultCommandLineActionFactoryTest.groovy
+++ b/platforms/core-runtime/gradle-cli/src/test/groovy/org/gradle/launcher/cli/DefaultCommandLineActionFactoryTest.groovy
@@ -116,19 +116,34 @@ class DefaultCommandLineActionFactoryTest extends Specification {
 
     def "reports command-line parse failure"() {
         when:
-        def commandLineExecution = factory.convert(['--broken'])
+        def commandLineExecution = factory.convert(options.split().toList())
         commandLineExecution.execute(executionListener)
 
         then:
-        outputs.stdErr.contains('--broken')
+        outputs.stdErr.contains(expectedMessage)
         outputs.stdErr.contains('USAGE: gradle [option...] [task...]')
         outputs.stdErr.contains('--help')
+        !outputs.stdErr.contains('To see help contextual to the project, use gradle help')
+        outputs.stdErr.contains('To see more detail about a task, run gradle help --task <task>')
+        if (expectedSuggestion != null) {
+            assert outputs.stdErr.contains("For example, gradle help --task $expectedSuggestion")
+        } else {
+            assert !outputs.stdErr.contains("For example, gradle help --task ")
+        }
+        outputs.stdErr.contains('To see a list of available tasks, run gradle tasks')
         outputs.stdErr.contains('--some-option')
 
         and:
         1 * actionFactory1.configureCommandLineParser(!null) >> {CommandLineParser parser -> parser.option('some-option')}
         1 * executionListener.onFailure({it instanceof CommandLineArgumentException})
         0 * executionListener._
+
+        where:
+        options                   | expectedMessage                                             | expectedSuggestion
+        '--broken'                | "Unknown command-line option '--broken'"                    | null
+        'wrapper --version=1.2.3' | "Command-line option '--version' does not take an argument" | 'wrapper'
+        '--version=1.2.3'         | "Command-line option '--version' does not take an argument" | null
+        '--wrapper'               | "Unknown command-line option '--wrapper'"                   | null
     }
 
     def "reports failure to build action due to command-line parse failure"() {
@@ -191,6 +206,9 @@ class DefaultCommandLineActionFactoryTest extends Specification {
 
         then:
         outputs.stdOut.contains('To see help contextual to the project, use gradle help')
+        outputs.stdOut.contains('To see more detail about a task, run gradle help --task <task>')
+        !outputs.stdOut.contains("For example, gradle help --task")
+        outputs.stdOut.contains('To see a list of available tasks, run gradle tasks')
         outputs.stdOut.contains('USAGE: gradle [option...] [task...]')
         outputs.stdOut.contains('--help')
         outputs.stdOut.contains('--some-option')
@@ -201,6 +219,28 @@ class DefaultCommandLineActionFactoryTest extends Specification {
 
         where:
         option << ['-h', '-?', '--help']
+    }
+
+    def "displays example for task "() {
+        when:
+        def commandLineExecution = factory.convert(options.split().toList())
+        commandLineExecution.execute(executionListener)
+
+        then:
+        outputs.stdOut.contains('To see help contextual to the project, use gradle help')
+        outputs.stdOut.contains('To see more detail about a task, run gradle help --task <task>')
+        outputs.stdOut.contains("For example, gradle help --task init")
+        outputs.stdOut.contains('To see a list of available tasks, run gradle tasks')
+        outputs.stdOut.contains('USAGE: gradle [option...] [task...]')
+        outputs.stdOut.contains('--help')
+        outputs.stdOut.contains('--some-option')
+
+        and:
+        1 * actionFactory1.configureCommandLineParser(!null) >> { CommandLineParser parser -> parser.option('some-option') }
+        0 * executionListener._
+
+        where:
+        options << ['-h init', 'init -?', 'init --help', '--help init']
     }
 
     def "uses system property for application name"() {

